src/lib/gameService.ts

```diff
@@ -12,41 +13,172 @@ export class GameService {
 			code: this.generateCode(),
 			currentPlayer: undefined,
 			players: [],
-			currentBid: undefined
+			currentBid: undefined,
+			maxPlayers: 8,
+			numberOfPlayers: 0,
+			initialDiceCount: 6
 		};
 
 		this.repository.saveGame(game);
 
 		return game.code;
 	}
 
-	public async addPlayer(name: string, gameCode: string): Promise<string> {
-		throw new Error('not implemented');
+	public async addPlayer(playerName: string, gameCode: string): Promise<string> {
+		const game: Game | undefined = await gameRepository.getGame(gameCode);
+
+		if (!game) {
+			throw new Error('Game is undefined');
+		}
+
+		if (game.state === GameState.InProgress) {
+			throw new Error('The game is already in progress');
+		} else if (game.state === GameState.Finished) {
+			throw new Error('The game has already finished');
+		}
+
+		const newPlayer: Player = {
+			name: playerName,
+			code: gameCode,
+			dice: []
+		};
+
+		if (game.players.length <= game.maxPlayers) {
+			game.players.push(newPlayer);
+		} else {
+			throw new Error('max player cap check failed');
+		}
+
+		this.repository.saveGame(game);
+
+		return playerName;
+
+		// throw new Error('not implemented');
 	}
 
-	public async getPlayers(name: string, gameCode: string): Promise<PlayerDto[]> {
-		throw new Error('not implemented');
+	// Should this instead return OpponentDto[]?
+	public async getPlayers(gameCode: string): Promise<PlayerDto[]> {
+		const game: Game | undefined = await gameRepository.getGame(gameCode);
+
+		if (!game) {
+			throw new Error('Game is undefined');
+		}
+
+		return game.players;
+
+		// throw new Error('not implemented');
 	}
 
-	public async startGame(gameCode: string): Promise<void> {
-		throw new Error('not implemented');
+	public async startGame(playerToken: string): Promise<void> {
+		const gameCode = this.getGameCode(playerToken);
+		const game: Game | undefined = await gameRepository.getGame(gameCode);
+		if (!game) {
+			throw new Error('Game is undefined');
+		}
+
+		if (game.players[0].name !== this.getPlayerName(playerToken)) {
+			throw new Error('Player is not the host');
+		}
+
+		const diceRoller: Roller = new Roller();
+		const numPlayers = game.players.length;
+
+		game.players.forEach((player) => {
+			player.dice = diceRoller.rollDice(game.initialDiceCount);
+		});
+		// ^^^
+		// |||  What do you think of making the dice roller roll all of the dice in the game in
+		// its "rollDice" function because you would always roll all of the dice.
```

Comment:
Good point that all the dice need to be rolled at once, that hadn't dawned on me.

I think that the roller should be kept dumb and not know what a game is - how many dice each player rolls is game logic and belongs in the game service imo.

The main reason for having a separate roller class is for all of the random elements of the game to sit outside of the game service so the game service is easier to test.